Support HTTPS connections to Huawei LTE devices

scop opened the pull request:
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->

Not all devices support HTTPS, so we default to plain HTTP still.

Ones that do are very likely to have certificates that do not pass hostname verification, and are either self signed or issued by an untrusted CA. Add option to use unverified HTTPS to make it possible to connect to these, and when in effect, filter urllib3's related warnings about insecure connections to the hostname in question.

This is related to #85875, but that one is not a prerequisite. We can target the warnings filter somewhat better than we could do the logging filter.

Requires/stacks on top of PR #85929.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: #85929 #85875
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

scop committed:
Support HTTPS connections to Huawei LTE devices

Not all devices support HTTPS, so we default to plain HTTP still.

Ones that do are very likely to have certificates that do not pass
hostname verification, and are either self signed or issued by an
untrusted CA. Add option to use unverified HTTPS to make it possible to
connect to these, and when in effect, filter urllib3's related warnings
about insecure connections to the hostname in question.

scop commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -58,6 +58,7 @@
     ATTR_CONFIG_ENTRY_ID,
     CONF_MANUFACTURER,
     CONF_UNAUTHENTICATED_MODE,
+    CONF_UNVERIFIED_HTTPS,
```
Maybe `homeassistant.const.CONF_VERIFY_SSL` should be used instead of this :thinking:

scop commented on homeassistant/components/huawei_lte/strings.json:
```diff
@@ -27,6 +29,7 @@
       "user": {
         "data": {
           "password": "[%key:common::config_flow::data::password%]",
+          "unverified_https": "Use unverified HTTPS",
```
Maybe we should use `[%key:common::config_flow::data::verify_ssl%]` here instead and adjust referring strings accordingly.

"Verify SSL certificate" is somewhat unfortunate phrasing though, in practice it's TLS we do nowadays, but maybe the general public is better accustomed to SSL than that. Anyway HTTPS would fit better as a term here, so in that sense we might want to keep this phrasing. Thoghts welcome.

emontnemery replied:
The existing language is used by 20 integrations, so it probably makes sense to use the same here?
Could the existing phrase be adjusted to something like "Verify SSL/TLS certificate"?

scop replied:
Yeah, I'll follow suit here, 9e2060ae8cbee05e0ad3be3c64e4372231e93e26 switches to the existing config and string keys.

We could see what others think about adjusting the existing phrase, but I think it's off topic for this PR. "Verify TLS/SSL certificate, "Verify TLS certificate", "Verify SSL/TLS certificate", "Verify certificate" could be candidates worth considering. ("Verify HTTPS certificate" would be nice, but the key might be already in use for other TLS verifications besides HTTPS.)

scop replied:
BTW I would prefer if we'd merge #85929 before this instead of this to its branch in order to not unnecessarily balloon it. Kind of would make sense to mark this as draft for that reason, but it's really not a draft but all ready for review, so I'd rather not.

scop replied:
#85929 has been merged, this one rebased and good to go.

scop committed:
Use common config key and strings for certificate verification settings

Even though the wording might be slightly suboptimal here, it's better
to be consistent across the codebase than to finetune on this level.

This also switches the default the other way around: verification is
now disabled by default. This is not a good general default, but for
this particular case setups where the verification would succeed would
be so rare and require considerable local setup that it's very
unlikely to happen in practice.

scop committed:
Add config flow tests

scop committed:
Mock logout for better test coverage

edenhaus commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -80,6 +81,13 @@ async def _async_show_user_form(
                             self.context.get(CONF_URL, ""),
                         ),
                     ): str,
+                    vol.Optional(
```
I suggest adding this option into a separate step as it could confuse the users if they insert an HTTP URL.

Maybe even better would be that we try the first time with validation activated, and if it fails, show a step where the user can turn it off.

scop replied:
I'm a somewhat torn on this one. I do think the suggestions make sense, but:

We're essentially doing the second already here: if a verified HTTPS connection fails due to a verification error, we say so and give a hint to retry with verification disabled. I'm not sure if a second separate step would be any more helpful.
Verification is off by default because I'm not aware of a router that would/could have a generally out of the box trusted cert installed, so verification must have been explicitly toggled on by the user, and as such, error should not come as a surprise.

With regards to the confusion with a plain HTTP URL, I'm not sure if that's a big issue. We could amend the help text with something like " (if https connection specified)" to alleviate, but that would mean giving up on using the common string for `verify_ssl`, and thus being possibly a bit inconsistent with other things in the future. (I see `vol.Optional` has a `description`, but it doesn't seem to appear in the UI unless I've missed something.)

edenhaus commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -119,11 +127,17 @@ async def _connect(
+            if user_input.get(CONF_VERIFY_SSL):
+                requests_session = None
+            else:
+                requests_session = non_verifying_requests_session(user_input[CONF_URL])
```
I suggest only to modify the session if the URL starts with `https://` and verification is disabled.

scop replied:
Fair enough, done in 4cb0104b993be3a5adc947bf4e1dee030875b2a0, also in `__init__.py`.

scop committed:
Set up custom requests session only when using unverified https

emontnemery reviewed (changes requested):
Please make sure the coverage requirement for the config flow is met

scop committed:
Merge branch 'dev' into scop-huawei-lte-https

scop committed:
Add https config flow test case

emontnemery commented on homeassistant/components/huawei_lte/config_flow.py:
```diff
@@ -119,11 +127,19 @@ async def _connect(
         password = user_input.get(CONF_PASSWORD) or ""
 
         def _get_connection() -> Connection:
+            if user_input[CONF_URL].startswith("https://") and not user_input.get(
```
`CONF_VERIFY_SSL` has a default value, why do we use `get` here?

scop replied:
d25b22448e818c4b8951fff1ac7d479967b0e3a8

emontnemery commented on homeassistant/components/huawei_lte/__init__.py:
```diff
@@ -335,16 +336,19 @@ async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+        kwargs: dict[str, Any] = {
+            "timeout": CONNECTION_TIMEOUT,
+        }
+        if url.startswith("https://") and not entry.data.get(CONF_VERIFY_SSL):
```
Before this PR, https was not supported.
With this PR, config entry data will always have `CONF_VERIFY_SSL` set.

Hence, can why do we use `get`?

scop replied:
Good catch, addressed in d25b22448e818c4b8951fff1ac7d479967b0e3a8.

emontnemery reviewed (changes requested):
Looks good, but please consider not using `get`.

scop committed:
Make better use of verify SSL default

emontnemery reviewed (approved):
LGTM, thanks @scop 👍

scop commented:
Thanks!
